### PR TITLE
Tooltip

### LINKED ISSUE #114: Add Tooltip Component
Tabler uses tooltips for a number of items.
For example, the media component uses tooltips:
screen shot 2018-04-28 at 17 10 27

Since we are trying to get rid of bootstrap components, a component could be developed from using [react-popper](https://github.com/FezVrasta/react-popper).

### PR DESCRIPTION
**Pull request by MrJarv1s**
This PR resolves #114 

The new tooltip use the original Tabler.css classes in addition to a new stylesheet included with the component.

fulltooltip

Before merging, these need to be completed:
- [ ] Add placement-based arrow positioning

### CODE REVIEW

File: src/components/Tooltip/Tooltip.react.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            <div
+              ref={ref}
+              onMouseEnter={this._handleTriggerOnMouseEnter}
+              onMouseLeave={this._handleTriggerOnMouseLeave}
+            >
+              {children}
+            </div>
+          )}
+        </Reference>
+        {this.state.isShown ? (
```
Comment (AaronCoplan) [inline src/components/Tooltip/Tooltip.react.js]: You can do `{this.state.isShown && (` and then you can leave out the `: null`

File: src/components/Tooltip/tooltip.css [hunk 2]
```diff
@@ -0,0 +1,5 @@
+/* Tooltip-specific Stylesheet */
```
Comment (AaronCoplan) [inline src/components/Tooltip/tooltip.css]: [nit] Let's capitalize the filename to Tooltip.css

File: src/components/Tooltip/Tooltip.react.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    e.preventDefault();
+    this.setState({ isShown: false });
+  };
+
+  render(): React.Node {
+    const { className, children, placement, content } = this.props;
+
+    const classes = cn(
+      "tooltip",
+      placement ? "bs-tooltip-" + placement : null,
```
Comment (AaronCoplan) [inline src/components/Tooltip/Tooltip.react.js]: Not sure, but you may be able to just do `placement && "bs-tooltip-" + placement,`

**Comment by MrJarv1s** [comment]
Wait my renaming of Tooltip.css did not show up... give a few minutes to fix that

**Comment by MrJarv1s** [comment]
Done!

**Comment by AaronCoplan** [comment]
Looks great to me!  I think the only missing piece would be to add it to the autogenerated documentation!  @jonthomp can probably help you out with that if you need it

**Comment by MrJarv1s** [comment]
@jonthomp I’d love to know about the automatic doc, brief tutorial on the subject?

**Comment by AaronCoplan** [comment]
Off the top of my head I think the steps are:

1. create a Tooltip.examples.md file in the same dir as Tooltip.react.js
2. Add it to the doc config
3. Run the script (look in package.json) to update the doc
4. Check and make sure it works by looking at the documentation page locally
5. Commit

**Comment by MrJarv1s** [comment]
Will do

**Comment by jonthomp** [comment]
@MrJarv1s couple of things to note, in the XXX.examples.md files your jsx code blocks automatically turn into the working examples with code preview that you see in the docs, comments above props in your flow prop type definitions are also pulled through to the docs automatically e.g.

```jsx
type Props {
	/**
	* This prop does something
	*/
	aprop: string,
	/**
	* This prop does something else
	*/
	bprop: number
}
```

**Comment by MrJarv1s** [comment]
Just finished writting the docs, ran `yarn styleguide` and checked my Tooltip. 

👎 

Apparently, the `<div>` block created around the reference element does not work for all components. It causes the tooltip to be misaligned in certain scenarios (Tags have this problem). It probably affects some other component types as well. I'll have to investigate a bit more. See the following picture for an example:

screen shot 2018-05-27 at 22 07 50

The code used for this example is the following:

```jsx
<Tooltip content="Tooltip" placement="top">
    <Tag>Hover Me!</Tag>
</Tooltip>
```

I think the `<div>` might be the issue, I had originally used it to create a new box that had the `onMouseEnter` and `onMouseLeave` props. Now that it might cause a problem with some other components, I'll see if I can use `React.cloneElement` to provide the props directly to the component.

Any ideas?

Also, why is my new local stylesheet for Tooltip not being used in the doc?

**Comment by jonthomp** [comment]
What was the issue with using React.cloneElement? Could it be solved with stricter child types, forwarding the refs to deeper child components and/or perhaps providing 2 routes in for users, children and then a render function passing them the ref and functions to use themselves?

On that stylesheet issue, sounds like the same as #170. Basically, our docs are built using CRAs Webpack config, but it doesn't support CSS modules. Our library is using rollup for its build, which does. Tere is a clash of the 2 because our docs are generated from CSS modules supporting code but built with Webpack. CRA 2 should fix this, I might get a branch up and running to test

**Comment by MrJarv1s** [comment]
From my first test, simple components such as `<a>` had a tooltip, where as `<Tab>` did not anymore. I also tried on `<Avatar>` and it did not work either.

I think your ref forwarding could work, I’ll have to test that out tonight.

**Comment by jonthomp** [comment]
Ah yes I had a similar issue with Dropdown, ended up adding a `rootRef` prop to Button and Dropdown.Trigger. I guess the only downside is that we'll need add rootRef props to all of our components that users might use inside a Tooltip.

**Comment by MrJarv1s** [comment]
I’ll test `rootRef` with the Tag component, if it works I’ll make a list of all components that could have this applied to. I might be in for a lengthy refactor, but I really want tooltip to be user-friendly.

**Comment by MrJarv1s** [comment]
Ok so there might be more than just the `rootRef`, since the tooltip uses `onMouseEnter` and `onMouseLeave` these must be included as props (used by components in their render).

For example, tooltips work well with `<a>` right now because it can add the prop directly to it (most likely because the prop is already implemented in React's `<a>`). For our components, we do not have these event props (or any other for that matter on most of the components). Because we do not make use of the event props in the render methods, it doesn't show up. 

I came to this conclusion while testing out the `rootRef` on the `Button` component and it did not work. I then added the mouse events to the `Button` component and it worked, see below:

screen shot 2018-05-28 at 12 24 33

I will start a list of all components requiring the update with `rootRef`, `onMouseEnter` and `onMouseLeave`. Are there any other events we would like added to all components? I could potentially add them as part of this PR.

**Comment by MrJarv1s** [comment]
Here are a list of base components, which I believe could use the `rootRef`, `onMouseEnter`, `onMouseLeave` props:

- Avatar
- Badge
- Button
- Form (most of them)
- Header
- Icon
- Tag
- Text

Most if not all components should have all possible mouse events as props, this would allow users more flexibility in their scripting. Will create an issue and link it.

**Comment by MrJarv1s** [comment]
Agreed, I will post the documentation update and get on the other thread.

**Comment by MrJarv1s** [comment]
@AaronCoplan, have you had the chance to review?

**Comment by AaronCoplan** [comment]
@MrJarv1s Sorry for the delay!  Looks good - approved for merge.  Only thing potentially missing is that I noticed the documentation index.html file hasn't been modified - I think you need to run the command to build the docs before merging 👍

**Pull request merged**